### ISSUE TITLE
Deeplink redirect on Microsoft 365 after external OAuth  

### ISSUE THREAD

**Issue by gazben** [OfficeDev/microsoft-teams-library-js#1696]
During our app certification we received report, that our external OAuth solution is not working with the Microsoft 365 app.  With the [new extended platform](https://learn.microsoft.com/en-us/microsoftteams/platform/m365-apps/overview) our application should work the same way in the `Teams` and `Microsoft 365 (Office)` .

All the documentation that I could find used this deeplink template:

`teams://teams.microsoft.com/l/auth-callback?authId=$authId&result=$result"`

With this, the Office 365 application won't recognize the redirent and will open the wrong application. How to do a proper deeplink redirect with  `Microsoft 365 (Office)`  if we are using out own OAuth solution?

**Comment by sayali-MSFT** [comment]
@gazben - Thanks for reporting your issue.
We will check this at our end and will get back to you.

**Comment by gazben** [comment]
@sayali-MSFT Do you have any update on this?

**Comment by sayali-MSFT** [comment]
@gazben -Have you check the generated deeplink format?
`msteams://teams.microsoft.com/l/auth-callback?authId=${state.authId}&result=${req.query.code}`

Did you get any error in console?

Ref Doc- https://learn.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/authentication/auth-oauth-provider#add-authentication-to-external-browsers

**Comment by gazben** [comment]
@sayali-MSFT Yes, it works for the Teams application. But this will always open the Teams application, even if I initiated the login from the O365 application.

**Comment by sayali-MSFT** [comment]
@gazben -Thanks for your quick response.
We will check this with internal team if it is bydesign behaviour or is there any other way to open link in Microsoft 365 (Office)?

**Comment by gazben** [comment]
Thank you, please check it internally if this is intentional or not.

**Comment by alicialu-MSFT** [comment]
Hi @gazben , could you confirm whether you are trying to deploy on O365 on desktop or on browser (office.com)?

**Comment by gazben** [comment]
I want to use this where the redirect method is `deeplink`: on desktop applications and mobile.

On web/browser, the redirect is done by a http redirect to the authentication end page where I call the `authentication.notifySuccess()` function.

**Comment by gazben** [comment]
We will use the `isExternal=false` for now, but this breaks the Google SSO.

**Comment by alicialu-MSFT** [comment]
Hi @gazben , we're working on a fix for this. Documentation will be updated when this is supported, but for now unfortunately it will not redirect to O365 after Google OAuth. We'll keep you updated -- thanks for bringing this up!

**Comment by onmomo** [comment]
@alicialu-MSFT do you have any updates on this. This has also been highlighted by the teams submission team during app certification. We can't use isExternal: false as workaround.

**Comment by onmomo** [comment]
@alicialu-MSFT I think I figured it out myself. Please update the docs.
```
// ensure to provide {hostRedirectUrl}
authentication
      .authenticate({
        url: `${window.location.origin}/auth/start.html?oauthRedirectMethod={oauthRedirectMethod}&authId={authId}&hostRedirectUrl={hostRedirectUrl}`,
        width: 970,
        height: 870,
        isExternal: true // browser based login, required for Teams Mobile and Teams 2.0 Desktop to allow secure oAuth authentication
      })
      
// then use the url provided in `hostRedirectUrl` instead following the MS docs, which only works for teams clients
if (state.oauthRedirectMethod === 'deeplink') {
   // this ensures isExternal flow also works with iOS Outlook and iOS M365 clients and what ever will come I guess.
   const clientRedirectUrl = state.hostRedirectUrl.replace('{result}', req.query.code)
   return res.redirect(clientRedirectUrl)
} 
```

**Comment by ChetanSharma-msft** [comment]
Hello @onmomo - Thanks for raising your query.
We are checking this with engineering team and let you know the updates.

**Comment by onmomo** [comment]
@ChetanSharma-msft do you have any updates on this? As I understand, the MS Teams mobile app (iOS/Android) and the M365 office mobile app do not implement the external OAuth flow as required since [teams-js 2.11.0](https://github.com/MicrosoftDocs/msteams-docs/pull/9307/files#diff-26b3157a2497a1fe89adeb2b0d33d4511a6287de73177470fa9e109dea0b6328R56). The apps in question do not replace the parameter `hostRedirectUrl` with a url to redirect to the corresponding as expected and just provide the {placeholder}.

Please find this table that documents my findings. We are using `teams-js:2.19.0`, manifest version `1.13`.

| MS App (latest version) | supports hostRedirectUrl parameter |
|--------|--------|
| Teams Mobile iOS / Android | does not replace parameter |
| M365 Office Mobile iOS / (maybe Android) | does not replace parameter |
| Outlook Mobile iOS / Android | does support parameter | 
| Teams 2.0 MacOS Desktop | does support parameter |

This is blocking the teams validation of our apps, please advise.

**Comment by vikramtha** [comment]
@onmomo Hey, I am reading over your comments, thank you for the table. I'm guessing that in the first comment you had mistakenly mentioned Teams iOS/Office iOS in the wrong places. You mentioned initially that Teams iOS does replace hostRedirectUrl and then later said it doesn't. 

I'm guessing this table is the updated truth as it matches up with what I am understanding as well.

I'll go ahead and talk with some people to update the documentation and investigate more into the mobile not replacing hostRedirectUrl.

**Comment by onmomo** [comment]
@vikramtha yes correct, thanks for pointing that out. The misleading comment is now updated, the table represents what we observed so far in our logs. Looking forward to your findings, thank you! 🙏

**Comment by vikramtha** [comment]
@onmomo  Okay I reached out to the M365 iOS team and they will start fixing the bug. No current eta as of now. I'll follow up with Teams shortly

**Comment by onmomo** [comment]
> @onmomo Okay I reached out to the M365 iOS team and they will start fixing the bug. No current eta as of now. I'll follow up with Teams shortly

@vikramtha do you have any update on this? I'm closely watching the app updates but non of them resolved the issue yet. Maybe it would be possible to provide us with the app deep link url for MS365 mobile apps (e.g for outlook `ms-outlook://mos/auth-callback?authId=4A5A1721-D448-4605-8843-F92C9BFC0000&result={result}`) in the mean time then we could already implement a workaround until this is fixed in the app

**Comment by onmomo** [comment]
@vikramtha Just figured out that the Outlook (new) version 1.2024.319.300 (Production) also fails to support the required `hostRedirectUrl ` parameter, it just provides an empty value. I've updated the table.

**Comment by vikramtha** [comment]
@onmomo  Thanks for the heads-up just communicated this with Outlook dev. The Teams Mobile folk btw probably will take a while to update this btw

**Comment by vikramtha** [comment]
@onmomo  Sorry but I followed up with the new Outlook client and unfortunately they do not have an ETA to fix any of this. They are currently working on improving the NAA stack.  Since the Outlook Win32 experience works you may want to default to this in the meantime.

**Comment by gercobrandwijk** [comment]
> > @onmomo Okay I reached out to the M365 iOS team and they will start fixing the bug. No current eta as of now. I'll follow up with Teams shortly
> 
> @vikramtha do you have any update on this? I'm closely watching the app updates but non of them resolved the issue yet. Maybe it would be possible to provide us with the app deep link url for MS365 mobile apps (e.g for outlook `ms-outlook://mos/auth-callback?authId=4A5A1721-D448-4605-8843-F92C9BFC0000&result={result}`) in the mean time then we could already implement a workaround until this is fixed in the app

Did you already manage to implement a workaround for the platforms where it currently does not work? 
For the new outlook it is `ms-outlook://mos/auth-callback?authId=4A5A1721-D448-4605-8843-F92C9BFC0000&result={result}`? What about the mobile apps, these should be deeplinks paths?

**Comment by onmomo** [comment]
> @ChetanSharma-msft do you have any updates on this? As I understand, the MS Teams mobile app (iOS/Android) and the M365 office mobile app do not implement the external OAuth flow as required since [teams-js 2.11.0](https://github.com/MicrosoftDocs/msteams-docs/pull/9307/files#diff-26b3157a2497a1fe89adeb2b0d33d4511a6287de73177470fa9e109dea0b6328R56). The apps in question do not replace the parameter `hostRedirectUrl` with a url to redirect to the corresponding client and just provide the {placeholder}.
> 
> Please find this table that documents my findings. We are using `teams-js:2.19.0`, manifest version `1.13`.
> 
> MS App (latest version)	supports hostRedirectUrl parameter
> Teams Mobile iOS / Android	does not replace parameter
> M365 Office Mobile iOS / (maybe Android)	does not replace parameter
> Outlook Mobile iOS / Android	does support parameter
> Teams 2.0 MacOS Desktop	does support parameter
> Outlook Windows 11 Desktop ( 1.2024.319.300 (Production)	does not support parameter
> This is blocking the teams validation of our apps, please advise.

@vikramtha do you have any updates on this? I just checked again and the situation changed a little (M365) but it is still not possible to implement isExternal flow with all required Microsoft mobile apps

| MS App (latest version) | supports hostRedirectUrl parameter |
|--------|--------|
| Teams Mobile iOS / Android | does not replace parameter |
| M365 Office Mobile iOS / (maybe Android) | does provide deepLink now but Safari iOS reports "Safari cannot open the page because the address is invalid" -> `msoffice://office.com/l/auth-callback?authId=27E8DAF0-1554-40FB-97E7-D8D123B36A77&result=$code`  |
| Outlook Mobile iOS / Android | does support parameter | 
| Teams 2.0 MacOS Desktop | does support parameter |
| Outlook Windows 11 Desktop ( 1.2024.319.300 (Production) | does not support parameter |

We will run a new publish round soonish. Considering this is now open for several months, we would appreciate if MSFT could speed up things a litte, thank you!

**Comment by vikramtha** [comment]
@onmomo and @gercobrandwijk 

I have followed up with the Office Mobile team again as they attempted to fix this but it seems to be still an issue. Waiting on their ETA. I have also followed up with Teams mobile however we have the hardcoded URL in this situation. In terms of the new Outlook experience this is still not on par with the Outlook Win32 experience and so we are still waiting to see when it is possible to fix this issue

**Comment by gercobrandwijk** [comment]
What is the timeline of the full deployment of the new Outlook experience to all the customers? More and more of our customers are reporting issues that our app does not work within the new Outlook. Will the new Outlook experience not be released till all the Outlook Win32 features are ported?

**Comment by onmomo** [comment]
@vikramtha do you have any update on this, can confirm that it is still not fixed with the latest M365 iOS app

**Comment by vikramtha** [comment]
@onmomo Small Update: We are still looking into it as we pushed the fix and redirectURL is being replaced properly but the redirect isn't happening. We will keep fixing this

**Comment by vikramtha** [comment]
@onmomo 

The change for the M365 iOS app is now merged and rolling out in our rings. It should be in dog food later this week and then on the monthly release cycle soon. Thank you for being patient as I know this took a long time to update!

**Comment by gercobrandwijk** [comment]
That's good new!

Any idea when it will be fixed in the New Outlook?

**Comment by vikramtha** [comment]
@gercobrandwijk  The change should be done end of the month so will be reflected early October. Sorry for the delay
